File: sdk/php/src/Providers/Anonbox.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * anonbox.net (CCC) 渠道实现
 *
 * GET /en/ 解析 HTML 提取邮箱与 token（inbox/secret），
 * GET /{inbox}/{secret}/ 返回 mbox 文本，按 "From " 分块解析每封邮件。
 */
final class Anonbox
{
    private const PAGE_URL = 'https://anonbox.net/en/';
    private const BASE = 'https://anonbox.net';

    /** @var array<string,string> */
    private const HEADERS = [
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept' => 'text/html,application/xhtml+xml',
    ];

    /**
     * 去除 HTML 标签并反转义常见实体。
     */
    private static function stripTags(string $html): string
    {
        $s = preg_replace('/<[^>]+>/', '', $html) ?? '';
        $s = str_replace(['&nbsp;', '&amp;', '&lt;', '&gt;'], [' ', '&', '<', '>'], $s);
        return trim($s);
    }

    /**
     * 简单 31 进制哈希，用于缺失 Message-ID 时生成稳定 ID（对齐 Python 实现）。
     */
    private static function simpleHash(string $s): string
    {
        $h = 0;
        $len = strlen($s);
        for ($i = 0; $i < $len; $i++) {
            $h = ($h * 31 + ord($s[$i])) & 0xFFFFFFFF;
        }
        $n = $h;
        if ($n === 0) {
            return '0';
        }
        $digits = '0123456789abcdefghijklmnopqrstuvwxyz';
        $out = '';
        while ($n > 0) {
            $out = $digits[$n % 36] . $out;
            $n = intdiv($n, 36);
        }
        return $out;
    }

    /**
     * 解析 /en/ 页面，返回 [email, token, expires]。
     *
     * @return array{0:string,1:string,2:?string}
     */
    private static function parseEnPage(string $html): array
    {
        if (
            preg_match(
                '#<a href="https://anonbox\.net/([a-z0-9-]+)/([A-Za-z0-9._~-]+)">https://anonbox\.net/[^"]+</a>#i',
                $html,
                $m,
            ) !== 1
        ) {
            throw new \RuntimeException('anonbox: mailbox link not found');
        }
        $inbox = $m[1];
        $secret = $m[2];
        $token = $inbox . '/' . $secret;

        $addressDisplay = null;
        if (preg_match_all('#<dd([^>]*)>([\s\S]*?)</dd>#i', $html, $dds, PREG_SET_ORDER)) {
            $expectedDomain = strtolower($inbox . '.anonbox.net');
            foreach ($dds as $dd) {
                if (preg_match('/display\s*:\s*none/i', $dd[1]) === 1) {
                    continue;
                }
                if (preg_match('#<p>([\s\S]*?)</p>#i', $dd[2], $pm) !== 1) {
                    continue;
                }
                $pInner = preg_replace('#<span\b[^>]*>[\s\S]*?</span>#i', '', $pm[1]) ?? $pm[1];
                $display = self::stripTags($pInner);
                if (!str_contains($display, '@')) {
                    continue;
                }
                $at = strrpos($display, '@');
                $local = trim(substr($display, 0, $at));
                $domain = strtolower(trim(substr($display, $at + 1)));
                if ($domain === 'googlemail.com' || $domain !== $expectedDomain || $local === '') {
                    continue;
                }
                $addressDisplay = $display;
                break;
            }
        }
        if ($addressDisplay === null) {
            throw new \RuntimeException('anonbox: address paragraph not found');
        }
        $at = strpos($addressDisplay, '@');
        $local = trim(substr($addressDisplay, 0, $at));
        if ($local === '') {
            throw new \RuntimeException('anonbox: empty local part');
        }
        $email = $local . '@' . $inbox . '.anonbox.net';

        $exp = null;
        if (preg_match('#Your mail address is valid until:</dt>\s*<dd><p>([^<]+)</p>#is', $html, $em) === 1) {
            $exp = trim($em[1]);
        }
        return [$email, $token, $exp];
    }

    /**
     * 按需解码 quoted-printable 正文。
     */
    private static function decodeQpIfNeeded(string $body, string $headerBlock): string
    {
        $enc = '';
        if (preg_match('/^content-transfer-encoding:\s*([^\s]+)/im', $headerBlock, $m) === 1) {
            $enc = strtolower(trim($m[1]));
        }
        if ($enc !== 'quoted-printable') {
            return rtrim($body, "\r\n");
        }
        return rtrim(quoted_printable_decode($body), "\r\n");
    }

    /**
     * 将单个 mbox 块解析为标准邮件原始字段。
     *
     * @return array<mixed>
     */
    private static function mboxBlockToRaw(string $block, string $recipient): array
    {
        $normalized = str_replace("\r\n", "\n", $block);
        $lines = explode("\n", $normalized);
        $i = 0;
        if (!empty($lines) && str_starts_with($lines[0], 'From ')) {
            $i = 1;
        }
        $headers = [];
        $curKey = '';
        $n = count($lines);
        while ($i < $n) {
            $line = $lines[$i];
            if ($line === '') {
                $i++;
                break;
            }
            if (($line[0] === ' ' || $line[0] === "\t") && $curKey !== '') {
                $headers[$curKey] .= ' ' . trim($line);
            } else {
                $idx = strpos($line, ':');
                if ($idx !== false && $idx > 0) {
                    $curKey = strtolower(trim(substr($line, 0, $idx)));
                    $headers[$curKey] = trim(substr($line, $idx + 1));
                }
            }
            $i++;
        }
        $body = implode("\n", array_slice($lines, $i));
        $ct = strtolower($headers['content-type'] ?? '');
        $text = '';
        $html = '';
        if (str_contains($ct, 'multipart/')) {
            if (preg_match('/boundary="?([^";\s]+)"?/i', $headers['content-type'] ?? '', $bm) === 1) {
                $boundary = preg_quote($bm[1], '#');
                $parts = preg_split('#\r?\n--' . $boundary . '(?:--)?\r?\n#', $body) ?: [];
                foreach ($parts as $part) {
                    $part = trim($part);
                    if ($part === '' || $part === '--') {
                        continue;
                    }
                    $sep = strpos($part, "\n\n");
                    if ($sep === false) {
                        continue;
                    }
                    $ph = substr($part, 0, $sep);
                    $pb = substr($part, $sep + 2);
                    $pct = '';
                    if (preg_match('/^content-type:\s*([^\s;]+)/im', $ph, $pctm) === 1) {
                        $pct = strtolower($pctm[1]);
                    }
                    if ($pct === 'text/plain') {
                        $text = self::decodeQpIfNeeded($pb, $ph);
                    } elseif ($pct === 'text/html') {
                        $html = self::decodeQpIfNeeded($pb, $ph);
                    }
                }
            }
        }
        if ($text === '' && $html === '') {
            if (str_contains($ct, 'text/html')) {
                $html = self::decodeQpIfNeeded($body, $headers['content-type'] ?? '');
            } else {
                $text = self::decodeQpIfNeeded($body, $headers['content-type'] ?? '');
            }
        }
        $dateStr = $headers['date'] ?? '';
        $msgId = $headers['message-id'] ?? self::simpleHash(substr($block, 0, 512));
        return [
            'id' => $msgId,
            'from' => $headers['from'] ?? '',
            'to' => $headers['to'] ?? $recipient,
            'subject' => $headers['subject'] ?? '',
            'text' => $text,
            'html' => $html,
            'date' => $dateStr,
            'isRead' => false,
            'attachments' => [],
        ];
    }

    public static function generate(): EmailInfo
    {
        $resp = HttpClient::get(self::PAGE_URL, self::HEADERS);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('anonbox: 首页请求失败 ' . $resp->getStatusCode());
        }
        [$email, $token, $exp] = self::parseEnPage((string) $resp->getBody());
        return new EmailInfo('anonbox', $email, $token, createdAt: $exp);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token): array
    {
        if ($token === null || $token === '') {
            throw new \InvalidArgumentException('anonbox: token 不能为空');
        }
        $path = str_ends_with($token, '/') ? $token : $token . '/';
        $resp = HttpClient::get(self::BASE . '/' . $path, [
            'User-Agent' => self::HEADERS['User-Agent'],
            'Accept' => 'text/plain,*/*',
        ]);
        if ($resp->getStatusCode() === 404) {
            return [];
        }
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('anonbox: 取信失败 ' . $resp->getStatusCode());
        }
        $raw = trim((string) $resp->getBody());
        if ($raw === '') {
            return [];
        }
        $blocks = preg_split('/\r?\n(?=From )/', $raw) ?: [];
        $out = [];
        foreach ($blocks as $b) {
            $b = trim($b);
            if ($b === '') {
                continue;
            }
            $out[] = Normalize::email(self::mboxBlockToRaw($b, $email), $email);
        }
        return $out;
    }
}

File: sdk/csharp/src/Providers/AnonBox.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// anonbox.net (CCC) 渠道。
/// 创建邮箱：GET /en/ 解析 HTML，提取邮箱地址与 inbox/secret 令牌。
/// 收信：GET /{inbox}/{secret}/ 返回 mbox 文本，按 "From " 行切分并解析每封邮件。
/// </summary>
public static class AnonBox
{
    private const string Channel = "anonbox";
    private const string PageUrl = "https://anonbox.net/en/";
    private const string BaseUrl = "https://anonbox.net";

    private const string Ua = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";

    private static readonly Regex MailLinkRe = new(
        "<a href=\"https://anonbox\\.net/([a-z0-9-]+)/([A-Za-z0-9._~-]+)\">https://anonbox\\.net/[^\"]+</a>",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex DdRe = new("<dd([^>]*)>([\\s\\S]*?)</dd>",
        RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);
    private static readonly Regex DisplayNoneRe = new("display\\s*:\\s*none", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex PRe = new("<p>([\\s\\S]*?)</p>", RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);
    private static readonly Regex SpanRe = new("<span\\b[^>]*>[\\s\\S]*?</span>", RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);
    private static readonly Regex TagRe = new("<[^>]+>", RegexOptions.Compiled);
    private static readonly Regex ExpiresRe = new("Your mail address is valid until:</dt>\\s*<dd><p>([^<]+)</p>",
        RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);
    private static readonly Regex MboxSplitRe = new("\\r?\\n(?=From )", RegexOptions.Compiled);

    private static string StripTags(string html)
    {
        var s = TagRe.Replace(html, "");
        return s.Replace("&nbsp;", " ").Replace("&amp;", "&").Replace("&lt;", "<").Replace("&gt;", ">").Trim();
    }

    /// <summary>与 Python _simple_hash 对齐的 31 进制哈希（base36 输出）</summary>
    private static string SimpleHash(string s)
    {
        uint h = 0;
        foreach (var c in s) h = (uint)((h * 31 + c) & 0xFFFFFFFF);
        if (h == 0) return "0";
        const string digits = "0123456789abcdefghijklmnopqrstuvwxyz";
        var sb = new StringBuilder();
        var n = h;
        while (n > 0) { sb.Insert(0, digits[(int)(n % 36)]); n /= 36; }
        return sb.ToString();
    }

    private static (string email, string token, string? exp) ParseEnPage(string html)
    {
        var m = MailLinkRe.Match(html);
        if (!m.Success) throw new Exception("anonbox: mailbox link not found");
        var inbox = m.Groups[1].Value;
        var secret = m.Groups[2].Value;
        var token = $"{inbox}/{secret}";

        string? addressHtml = null;
        foreach (Match dd in DdRe.Matches(html))
        {
            if (DisplayNoneRe.IsMatch(dd.Groups[1].Value)) continue;
            var pm = PRe.Match(dd.Groups[2].Value);
            if (!pm.Success) continue;
            var pInner = SpanRe.Replace(pm.Groups[1].Value, "");
            var display = StripTags(pInner);
            if (!display.Contains('@')) continue;
            var at2 = display.LastIndexOf('@');
            var domain = display[(at2 + 1)..].Trim().ToLowerInvariant();
            if (domain == "googlemail.com") continue;
            var expectedDomain = $"{inbox}.anonbox.net".ToLowerInvariant();
            if (domain != expectedDomain) continue;
            var local2 = display[..at2].Trim();
            if (local2.Length == 0) continue;
            addressHtml = display;
            break;
        }
        if (addressHtml is null) throw new Exception("anonbox: address paragraph not found");

        var at = addressHtml.IndexOf('@');
        if (at < 0) throw new Exception("anonbox: bad address");
        var local = addressHtml[..at].Trim();
        if (local.Length == 0) throw new Exception("anonbox: empty local part");
        var email = $"{local}@{inbox}.anonbox.net";

        var em = ExpiresRe.Match(html);
        var exp = em.Success ? em.Groups[1].Value.Trim() : null;
        return (email, token, exp);
    }

    private static string DecodeQpIfNeeded(string body, string headerBlock)
    {
        var te = Regex.Match(headerBlock, "^content-transfer-encoding:\\s*([^\\s]+)",
            RegexOptions.IgnoreCase | RegexOptions.Multiline);
        var enc = te.Success ? te.Groups[1].Value.ToLowerInvariant().Trim() : "";
        if (enc != "quoted-printable") return body.TrimEnd('\r', '\n');

        var soft = Regex.Replace(body, "=\\r?\\n", "");
        var b = Encoding.Latin1.GetBytes(soft);
        var outBytes = new List<byte>();
        var i = 0;
        while (i < b.Length)
        {
            if (b[i] == (byte)'=' && i + 2 < b.Length)
            {
                var hex = Encoding.ASCII.GetString(b, i + 1, 2);
                if (int.TryParse(hex, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var val))
                {
                    outBytes.Add((byte)val);
                    i += 3;
                    continue;
                }
            }
            outBytes.Add(b[i]);
            i++;
        }
        return Encoding.UTF8.GetString(outBytes.ToArray()).TrimEnd('\r', '\n');
    }

    private static Dictionary<string, object?> MboxBlockToRaw(string block, string recipient)
    {
        var normalized = block.Replace("\r\n", "\n");
        var lines = normalized.Split('\n');
        var i = 0;
        if (lines.Length > 0 && lines[0].StartsWith("From ")) i = 1;

        var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var curKey = "";
        while (i < lines.Length)
        {
            var line = lines[i];
            if (line == "") { i++; break; }
            if ((line[0] == ' ' || line[0] == '\t') && curKey.Length > 0)
            {
                headers[curKey] = headers[curKey] + " " + line.Trim();
            }
            else
            {
                var idx = line.IndexOf(':');
                if (idx > 0)
                {
                    curKey = line[..idx].Trim().ToLowerInvariant();
                    headers[curKey] = line[(idx + 1)..].Trim();
                }
            }
            i++;
        }
        var body = string.Join("\n", lines[i..]);

        var ct = (headers.GetValueOrDefault("content-type") ?? "").ToLowerInvariant();
        var text = "";
        var html = "";
        if (ct.Contains("multipart/"))
        {
            var bm = Regex.Match(headers.GetValueOrDefault("content-type") ?? "",
                "boundary=\"?([^\";\\s]+)\"?", RegexOptions.IgnoreCase);
            if (bm.Success)
            {
                var boundary = Regex.Escape(bm.Groups[1].Value);
                var partRe = new Regex($"\\r?\\n--{boundary}(?:--)?\\r?\\n");
                foreach (var partRaw in partRe.Split(body))
                {
                    var part = partRaw.Trim();
                    if (part.Length == 0 || part == "--") continue;
                    var sep = part.IndexOf("\n\n", StringComparison.Ordinal);
                    if (sep < 0) continue;
                    var ph = part[..sep];
                    var pb = part[(sep + 2)..];
                    var pctM = Regex.Match(ph, "^content-type:\\s*([^\\s;]+)", RegexOptions.IgnoreCase | RegexOptions.Multiline);
                    var pct = pctM.Success ? pctM.Groups[1].Value.ToLowerInvariant() : "";
                    if (pct == "text/plain") text = DecodeQpIfNeeded(pb, ph);
                    else if (pct == "text/html") html = DecodeQpIfNeeded(pb, ph);
                }
            }
        }
        if (text.Length == 0 && html.Length == 0)
        {
            if (ct.Contains("text/html")) html = DecodeQpIfNeeded(body, headers.GetValueOrDefault("content-type") ?? "");
            else text = DecodeQpIfNeeded(body, headers.GetValueOrDefault("content-type") ?? "");
        }

        var dateStr = headers.GetValueOrDefault("date") ?? "";
        if (dateStr.Length > 0 && DateTimeOffset.TryParse(dateStr.Trim(), CultureInfo.InvariantCulture,
                DateTimeStyles.RoundtripKind, out var dto))
            dateStr = dto.ToString("o", CultureInfo.InvariantCulture);
        if (dateStr.Length == 0)
            dateStr = DateTimeOffset.UtcNow.ToString("o", CultureInfo.InvariantCulture);

        var msgId = headers.GetValueOrDefault("message-id");
        if (string.IsNullOrEmpty(msgId))
            msgId = SimpleHash(block.Length > 512 ? block[..512] : block);

        return new Dictionary<string, object?>
        {
            ["id"] = msgId,
            ["from"] = headers.GetValueOrDefault("from") ?? "",
            ["to"] = string.IsNullOrEmpty(headers.GetValueOrDefault("to")) ? recipient : headers["to"],
            ["subject"] = headers.GetValueOrDefault("subject") ?? "",
            ["body_text"] = text,
            ["body_html"] = html,
            ["date"] = dateStr,
            ["isRead"] = false,
            ["attachments"] = new List<object?>(),
        };
    }

    /// <summary>创建临时邮箱</summary>
    public static EmailInfo Generate()
    {
        var resp = Http.Get(PageUrl, new Dictionary<string, string>
        {
            ["User-Agent"] = Ua,
            ["Accept"] = "text/html,application/xhtml+xml",
        }, 15);
        resp.EnsureSuccess();
        var (email, token, exp) = ParseEnPage(resp.Body);
        return new EmailInfo(Channel, email, token, createdAt: exp);
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string? token, string email)
    {
        if (string.IsNullOrEmpty(token)) throw new ArgumentException("token is required for anonbox channel");
        var path = token.EndsWith("/") ? token : token + "/";
        var url = $"{BaseUrl}/{path}";
        var resp = Http.Get(url, new Dictionary<string, string>
        {
            ["User-Agent"] = Ua,
            ["Accept"] = "text/plain,*/*",
        }, 15);
        if (resp.StatusCode == 404) return new List<Email>();
        resp.EnsureSuccess();
        var raw = resp.Body.Trim();
        var result = new List<Email>();
        if (raw.Length == 0) return result;
        foreach (var b in MboxSplitRe.Split(raw))
        {
            var block = b.Trim();
            if (block.Length == 0) continue;
            result.Add(Normalize.NormalizeEmail(MboxBlockToRaw(block, email), email));
        }
        return result;
    }
}
